### ISSUE TITLE
Bug: allowDefaultProject mishandled under eslint 9.12.0

### ISSUE THREAD

**Issue by LukeNotable** [typescript-eslint/typescript-eslint#10159]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I'm getting this error under `eslint@9.12.0`, which did not occur under `eslint@9.11.1`:

```
/wherever/vitest.config.ts
  0:0  error  Parsing error: /wherever/vitest.config.ts was not found by the project service. 
Consider either including it in the tsconfig.json or including it in allowDefaultProject
```

The pertinent config is:
```ts
  {
    files: ['*.ts', '{src,db}/**/*.ts'],
    extends: [...ts.configs.recommendedTypeChecked],
    languageOptions: {
      parserOptions: {
        projectService: {
          allowDefaultProject: ['vitest.config.ts'],
        },
        tsconfigRootDir: import.meta.dirname,
      },
    },
  },
  {
    files: ['db/**/*.ts'],
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.dirname + '/db',
      },
    },
  },
```

The problem disappears if _any_ of these are changed:
* eslint is downgraded to lower version
* the last block (with `files: ['db/**/*.ts']`) is removed
* the first block is changed to not overlap with the second (i.e., `files: ['*.ts', 'src/**/*.ts']`)

### Reproduction Repository Link

https://github.com/LukeNotable/eslint-9.12.0-allow-default-project-bug

### Repro Steps

1) clone the repo
2) `yarn install`
3) `yarn eslint .`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.9.0` |
| `@typescript-eslint/parser`            | `8.9.0` |
| `@typescript-eslint/scope-manager`     | `8.9.0` |
| `@typescript-eslint/typescript-estree` | `8.9.0` |
| `@typescript-eslint/type-utils`        | `8.9.0` |
| `@typescript-eslint/utils`             | `8.9.0` |
| `TypeScript`                           | `5.5.4` |
| `ESLint`                               | `9.12.0` |
| `node`                                 | `22.9.0` |

**Comment by bradzacher** [comment]
have you tried using `eslint --print-config` and/or `eslint --inspect-config` to see what ESLint resolves for your files?

Does it show different information between the two versions?

**Comment by LukeNotable** [comment]
Wow, what a cool tool! 

No, the `--inspect-config` yields the same results for both eslint versions.

<img width="933" alt="image" src="https://github.com/user-attachments/assets/37d132e3-9edd-4ee8-a8a1-79a8d0bde747">

**Comment by JoshuaKGoldberg** [comment]
🙌 hooray, mysterious code fixing!

**Comment by LukeNotable** [comment]
Actually, now that I try it one more time, the problem is back. Must be flaky. 🙁

**Comment by JoshuaKGoldberg** [comment]
It doesn't repro on my end at all 😦

**Comment by bradzacher** [comment]
Try using the other flag - `--print-config` which will dump the config for a specific file.
That way you can see exactly what ESLint is applying to the file that's broken.

**Comment by LukeNotable** [comment]
Well, `yarn eslint vitest.config.ts` succeeds, and I need `yarn eslint .` to reproduce the problem, but `yarn eslint --print-config .` yields nothing. Anyway, this is the output for the one file. 

```
yarn eslint --print-config vitest.config.ts

{
  "linterOptions": {
    "reportUnusedDisableDirectives": 1
  },
  "rules": {
    // lots of rules
  },
  "plugins": [
    "@",
    "@typescript-eslint:@typescript-eslint/eslint-plugin@8.10.0"
  ],
  "language": "@/js",
  "languageOptions": {
    "sourceType": "module",
    "ecmaVersion": "latest",
    "parser": "typescript-eslint/parser@8.10.0",
    "parserOptions": {
      "projectService": {
        "allowDefaultProject": {
          "0": "vitest.config.ts"
        }
      },
      "tsconfigRootDir": "/Users/luke/eslint-9.12.0-allow-default-project-bug"
    },
    "globals": {}
  }
}
```

**Comment by bradzacher** [comment]
@JoshuaKGoldberg considering this works with one file linted but breaks when more files are linted - you reckon this is due to having two separate `projectService` configs?

**Comment by LukeNotable** [comment]
It's non-deterministic. For me it fails ~80% of the time with `yarn eslint .` (4 files), or ~20% of the time with `yarn eslint vitest.config.ts db/bar.ts` (2 files).

**Comment by LukeNotable** [comment]
Actually, in CI runs I'm occasionally seeing the problem even under eslint 9.11.1. So apparently it's been there a while and is just more surfaced now.

**Comment by JoshuaKGoldberg** [comment]
Ok, yes, I do see this reproducing sometimes now. It didn't the first couple of times I tried. 🫠  Doing a bit of debugging with `DEBUG=*`...

When project resolution succeeds, some relevant log lines:

```plaintext
  typescript-eslint:typescript-estree:createProjectService Creating project service with: { defaultProject: 'tsconfig.json', allowDefaultProject: [ 'vitest.config.ts' ] } +0ms
  ...
  typescript-eslint:typescript-estree:tsserver:info     FileName: /Users/josh/repos/eslint-9.12.0-allow-default-project-bug/vitest.config.ts ProjectRootPath: /Users/josh/repos/eslint-9.12.0-allow-default-project-bug +0ms
  typescript-eslint:typescript-estree:tsserver:info             Projects: /dev/null/inferredProject1* +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Result from attempting to open client file: { configFileName: undefined, configFileErrors: undefined } +183ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Default project allowed path: true, based on config file: undefined +0ms
```

When it fails:

```plaintext
  typescript-eslint:typescript-estree:createProjectService Creating project service with: { defaultProject: 'tsconfig.json' } +0ms
  ...
  typescript-eslint:typescript-estree:tsserver:info     FileName: /Users/josh/repos/eslint-9.12.0-allow-default-project-bug/vitest.config.ts ProjectRootPath: /Users/josh/repos/eslint-9.12.0-allow-default-project-bug +0ms
  typescript-eslint:typescript-estree:tsserver:info             Projects: /dev/null/inferredProject1* +0ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Result from attempting to open client file: { configFileName: undefined, configFileErrors: undefined } +140ms
  typescript-eslint:typescript-estree:useProgramFromProjectService Default project allowed path: false, based on config file: undefined +0ms
```

Going up, I see that when `createParseSettings` is first called with `db/bar.ts`, we do see the error. But when it's first called with `vitest.config.ts` then we don't see the error. So, yes, it looks like it's nondeterministic based on which of the two config blocks happens to hit the file first.

A root reason for this is that we create a single project service once per lint run:
https://github.com/typescript-eslint/typescript-eslint/blob/8d35958494eb0c97cc36c3fc0844aea377542315/packages/typescript-estree/src/parseSettings/createParseSettings.ts#L110-L120

We would _love_ to have more intelligent project service creation, but can't because ESLint core doesn't tell parsers such as `@typescript-eslint/parser` (which is what calls to `@typescript-eslint/typescript-estree`) session information. We asked for this in https://github.com/eslint/rfcs/pull/102 but it was not accepted.

In the meantime, I'm not sure the best way to do this. Maybe we could, in `createParseSettings`:

1. Detect if `tsestreeOptions.projectService` has changed (like a deep equality check)
2. If they are different, depending on whether it's reasonable to have multiple project services in existence:
  * If it's ok: have a `Map` hashing project service settings to project services
  * If it's not ok: throw an error

cc @bradzacher @jakebailey - I suspect we'd have to do some experimentation to see how bad it is to have multiple project services floating around. My hunch is that, no, that's really bad for memory in larger projects and shouldn't be allowed.

**Comment by jakebailey** [comment]
I would be curious if TS 5.7's project loading changes would make this more well-behaved; might be worth a try. There have been improvements to project loading which should result in projects being more correctly/optimally loaded (and even fewer loads).

**Comment by gtbuchanan** [comment]
This problem occurs randomly for us as well. We use a separate project configuration for "root" files (typically config)

![Image](https://github.com/user-attachments/assets/76276c97-357a-4498-9fbe-e4b431e7a810)

#### eslint.config.mjs
```javascript
    {
      files: ['*.{cjs,cts,js,mjs,mts,ts}'],
      languageOptions: {
        parserOptions: {
          projectService: {
            allowDefaultProject: ['*.js', '*.mjs', '*.ts'],
            defaultProject: 'tsconfig.root.json',
          },
        },
      },
    },
```

#### tsconfig.root.json
```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "noEmit": true,
  },
  "extends": ["./typescript/library-base/tsconfig"],
  "include": ["*"]
}
```

#### typescript/library-base/tsconfig.json
```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "declaration": true,
    "esModuleInterop": true,
    "lib": [ "ESNext" ],
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "target": "ESNext"
  },
  "extends": ["@tsconfig/strictest/tsconfig.json"]
}
```

**Comment by JoshuaKGoldberg** [comment]
I haven't had time to investigate this recently and probably won't soon. Un-assigning myself. 😕

**Comment by gtbuchanan** [comment]
For what it's worth, we haven't encountered this error in a while (months?) on TypeScript 5.7.3 and typescript-eslint 8.32.0.

**Comment by kirkwaiblinger** [comment]
> For what it's worth, we haven't encountered this error in a while (months?) on TypeScript 5.7.3 and typescript-eslint 8.32.0.

I've just tested this out and I can still reproduce with those versions, as well as latest TS and typescript-eslint 🫤. 

It seems like a fresh install with fresh terminal is more likely to work, whereas subsequent runs will fail? But not 100% according to that pattern.
